### ISSUE TITLE
twisted.conch.client.agent.SSHAgentClient.getPublicKey returns a str, triggers deprecated behavior from twisted.conch.client.default.SSHUserAuthClient.getPublicKey

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3813]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3813|
|Type|defect|
|Created|2009-05-03 18:29:10Z|
|Branch|https://github.com/twisted/twisted/tree/conch-agent-deprecations-3813|

`twisted.conch.client.default.SSHUserAuthClient.getPublicKey` is supposed to return a `twisted.conch.ssh.keys.Key` instance now.  It calls `twisted.conch.client.agent.SSHAgentClient.getPublicKey` if the environment is configured with an agent, though, which returns a `str` instance.  This results in a deprecation warning.

Either `SSHUserAuthClient` should take responsibility for converting the `str` into a `Key` before returning, or the `SSHAgentClient` should be updated to deal with `Key` objects too.

`SSHUserAuthClient` also triggers deprecation warnings in another case: if there's no agent, it uses `twisted.conch.ssh.keys.getPublicKeyString` which is deprecated and _also_ returns a `str`, causing the same deprecation as happens in the with-agent case.

<details><summary>Searchable metadata</summary>

```
trac-id__3813 3813
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_conch_agent_deprecations_3813 branches/conch-agent-deprecations-3813
branch_author__z3p__therve z3p, therve
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1241375350000000 1241375350000000
changetime__1246522097000000 1246522097000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|z3p commented|
|-|-|

(In [26865]) Branching to 'conch-agent-deprecations-3813'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|z3p removed owner|
|-|-|

Ready for review in conch-agent-deprecations-3813.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to z3p|
|-|-|

 * 
```
-            blob = self.keyAgent.getPublicKey()
-            if blob:
-                return blob
+            key = self.keyAgent.getPublicKey()
+            if key is not None:
+                return key
```

This change is not tested.

 * 
```
+            return keys.Key.fromFile(file)
+        except keys.KeyError:
```

The except is not tested.

 * 
```
+                    return defer.succeed(keys.Key.fromFile(file, passphrase = p))
```

Spaces around "=" are unnecessary.

 * 
```
+        self.tmpdir = self.mktemp()
+        os.mkdir(self.tmpdir)
+        self.rsaFile = os.path.join(self.tmpdir, 'id_rsa')
+        file(self.rsaFile + '.pub', 'wb').write(keydata.publicRSA_openssh)
+        file(self.rsaFile, 'wb').write(keydata.privateRSA_openssh)
```

Please use the `FilePath` `API` to create the directory, and `FilePath.setContent` to create the files (write without close relies on the GC behavior).

 * 
```
+        file(self.rsaFile, 'wb').write(rsaPrivate.toString('openssh', passphrase))
```

Same thing here.

 * `test_getPrivateKeyWithPassphrase` is missing a docstring.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|z3p removed owner|
|-|-|

I believe I've addressed these comments in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-agent-deprecations-3813)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to z3p|
|-|-|

Thanks for the fixes.  Now I can 'conch cube.twistedmatrix.com' without seeing _any_ warnings or errors during login!  Woo!

    1. `twisted.conch.client.agent.SSHAgentClient.getPublicKey` is changing, but it has no docstring.  Can you add one?  The fact that it's a destructive operation is potentially confusing.  I would personally be especially grateful if such a docstring said who was supposed to call it and why.
    1. `twisted.conch.client.default.SSHUserAuthClient.getPrivateKey` is changing, but it has no docstring.  Can you add one?  The fact that it can call `reactor.stop()` is particularly interesting.
    1. The coding standard says just one underscore in test names.  So can you rename e.g. `test_agent_getPublicKey` to `test_agentGetPublicKey` and so on?
    1. ♥ getting rid of string comparisons in `except:` blocks.  Thank you for that.
    1. Please put two blank lines between each new test method, rather than just one, as per the "whitespace" section of the coding standard.
    1. We've been gradually shifting our test-docstring-writing style to be declarative rather than normative.  This isn't covered in the coding standard, but it should be.  This was inspired by [https://launchpad.net/testdoc jml's testdoc tool], in the hopes that one day Trial will do something like that.  For example, instead of saying:
        `L{SSHUserAuthClient} should be able to get a private key from a file, and return a Deferred called back with a private L{Key} object.`
       say something like:
        `L{SSHUserAuthClient.getPrivateKey} will load a private key from the first file described by its options' C{identitys} list, and return a L{Deferred} which fires with the corresponding private L{Key}.`
       The docstrings are fine as they are, but I find that this style is a lot nicer to read as a maintainer, because it documents the behavior that the test is verifying, rather than focusing on how the verification is done.
    1. I noticed the following during this review and I think they'd be good points for discussion or possibly separate tickets:
        1. I can't help but notice that there's a missing layer here, which the usage of `TestCase.patch` in these tests highlights; `_getPassword` is a secret implementation detail of `twisted.conch.client.default.SSHUserAuthClient` rather than a nice separate layer that the tests can pass as a (public, documented) parameter.  Is there a ticket for this?  If not, there should be.  It seems that if we're going to implement things like [#3829](https://github.com/twisted/twisted/issues/3829) in a generally useful manner, the "default" client has to be easily usable in a programmatic manner.  `twisted.conch.client.knownhosts.ConsoleUI` might be a good place to start with this.
        1. There's a bunch of duplication in the new tests.  They're okay as tests go, but it seems like the construction of the options and the keys and stuff could be factored out.  Maybe a good 'easy' ticket to get somebody else involved with conch?
        1. There are a couple of bareword `except:`s going away.  I'm wondering if there were any other bugs that was working around which might now be exposed again.  Test coverage looks all right though, so I don't _think_ so...

Address these as appropriate and land; I don't see anything here which should require a re-review.  But add those docstrings, for reals :).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

I think I've handled the first 6 points. I guess the last ones can deserve their own tickets.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

  1. Please update copyright dates on modified files.
  1. In, twisted/conch/client/default.py, in `SSHUserAuthClient.getPrivateKey`, when `keys.EncryptedKeyError` is caught, it is bound to the local `e`, but that local is never used (pyflakes warns about this now).
  1. In twisted/conch/test/test_default.py, two new tests do `public = client.getPublicKey()`.  These calls are just for the side-effect of setting up the proper internal state so that `getPrivateKey` will work later.  It's probably worth putting a comment in explaining this, and maybe dropping the local variable assignment (ie, just call the method).  pyflakes also warns about these unused locals now.

Otherwise looks good.  Please merge when the above points are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [27076]) Merge conch-agent-deprecations-3813

Authors: z3p, therve
Reviewers: exarkun, glyph
Fixes [#3813](https://github.com/twisted/twisted/issues/3813)

Remove usage of deprecated Key APIs from twisted.conch.client.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
